lib/ansible/modules/cloud/smartos/imgadm.py

```diff
@@ -0,0 +1,302 @@
+version_added: "2.3"
+author: Jasper Lievisse Adriaanse (@jasperla)
+options:
+    force:
+        required: false
+        choices: [ yes, no ]
+        description:
+        - Force a given operation (where supported by imgadm(1M)).
+    pool:
+        required: false
+        default: zones
+        description:
+        - zpool to import to or delete images from.
+    source:
+        required: false
+        description:
+        - URI for the image source.
+    state:
+        required: true
+        choices: [ present, absent, deleted, imported, updated, vacumed ]
+        description:
+        - State the object operated on should be in. C(imported) is an alias for
+          for C(present) and C(deleted) for C(absent). When set to C(vacumed)
+          and C(uuid) to C(*), it will remove all unused images.
+    type:
+        required: false
+        choices: [ imgapi, docker, dsapi ]
+        default: imgapi
+        description:
+        - Type for image sources.
+    uuid:
+        required: false
+        description:
+        - Image UUID. Can either be a full UUID or C(*) for all images.
+'''
+
+EXAMPLES = '''
+- name: Import an image
+  imgadm:
+    uuid: '70e3ae72-96b6-11e6-9056-9737fd4d0764'
+    state: imported
+
+- name: Delete an image
+  imgadm:
+    uuid: '70e3ae72-96b6-11e6-9056-9737fd4d0764'
+    state: deleted
+
+- name: Update all images
+  imgadm:
+    uuid: '*'
+    state: updated
+
+- name: Update a single image
+  imgadm:
+    uuid: '70e3ae72-96b6-11e6-9056-9737fd4d0764'
+    state: updated
+
+- name: Add a source
+  imgadm:
+    source: 'https://datasets.project-fifo.net'
+    state: present
+
+- name: Add a Docker source
+  imgadm:
+    source: 'https://docker.io'
+    type: docker
+    state: present
+
+- name: Remove a source
+  imgadm:
+    source: 'https://docker.io'
+    state: absent
+'''
+
+import re
+
+# Shortcut for the imgadm(1M) command. While imgadm(1M) supports a
+# -E option to return any errors in JSON, the generated JSON does not play well
+# with the JSON parsers of Python. The returned message contains '\n' as part of
+# the stacktrace, which breaks the parsers.
+IMGADM = 'imgadm'
+
+# Helper method to massage stderr
+def errmsg(stderr):
+    match = re.match('^imgadm .*?: error \(\w+\): (.*): .*', stderr)
+    if match:
+        return match.groups()[0]
+    else:
+        return 'Unexpected failure'
+
+def update_images(module):
+    uuid = module.params['uuid']
+    cmd = IMGADM + ' update'
+
+    if uuid != '*':
+        cmd += ' {}'.format(uuid)
+
+    (rc, stdout, stderr) = module.run_command(cmd)
+
+    # There is no feedback from imgadm(1M) to determine if anything
+    # was actually changed. So treat this as an 'always-changes' operation.
+    # Note that 'imgadm -v' produces unparseable JSON...
+    return rc, stdout, errmsg(stderr), True
+
+def manage_sources(module, present):
+    force = module.params['force']
+    source = module.params['source']
+    type = module.params['type']
```

Comment:
Note (not a blocker): type overrides the builtn "type".  It's better to use a different name even if it's ```type_```.